Auterion Skynode upload scripts

MaEtUgR opened the pull request:
### Solved Problem
When building firmware for Auterion Skynode with `make px4_fmu-v5x` there is no upload script to directly flash it but the user has to manually flash the binary.

### Solution
- Add scripts to simplify flashing Skynode over either USB or Wifi.

### Changelog Entry
```
Tools: Skynode flash scripts
Documentation: Still needs to be added to the documentation
```

### Alternatives
Instead of `make px4_fmu-v5x upload_usb` it could be either a more specific upload target `make px4_fmu-v5x upload_skynode_usb` or a completely different build target `make auterion_skynode upload`. I'd need to check how the latter can be achieved without duplicating anything from the v5x target.

### Test coverage
- I did not test this pr but am using the same exact rebased scripts a lot. I can quickly double-check.

### Context
See example output in https://github.com/PX4/PX4-Autopilot/pull/21842#issuecomment-1634477701

hamishwillee commented on boards/px4/fmu-v5x/cmake/upload.cmake:
```diff
@@ -0,0 +1,49 @@
+set(PX4_FW_NAME ${PX4_BINARY_DIR}/${PX4_BOARD_VENDOR}_${PX4_BOARD_MODEL}_${PX4_BOARD_LABEL}.px4)
+
+add_custom_target(upload_usb
+	COMMAND ${PX4_SOURCE_DIR}/Tools/auterion/upload_skynode.sh ${PX4_FW_NAME} --default-ip=10.41.1.1
```
Would this IP address every change?

MaEtUgR replied:
I honestly don't know. It is like this since the beginning but I can't guarantee it will always stay the same in the future.

hamishwillee replied:
Then we document as defaults and modify to be configurable when they have to be.

MaEtUgR commented:
> Can the upload scripts for wifi and usb be PX4-generic

I thought about that and in principle yes but it's a bit further off because there is no other board currently implementing anything similar, to use the same script it would need to do it the same way and there's also no open, standardized software for the client/companion side to execute on that. So my conclusion is: In the current form no.

> but at the same time it doesn't make sense to have a specific skynode target for normal PX4 binaries

That's needed since flashing Skynode also over USB is through ssh which is technically different from the pixhawk USB flashing. That's why an upload script is needed at all and other than being called makes it work the same way for the user. We could in principle add auto-detection but that adds more complexity and failure cases.

hamishwillee commented:
> but at the same time it doesn't make sense to have a specific skynode target for normal PX4 binaries

> That's needed since flashing Skynode also over USB is through ssh which is technically different from the pixhawk USB flashing. That's why an upload script is needed at all and other than being called makes it work the same way for the user. We could in principle add auto-detection but that adds more complexity and failure cases.

Sorry, I was imprecise. I mean that we should use `px4_fmu-v5x` as the target, not `auterion_skynode`. Main reason being to support same upload command for targets on new skynode hardware.

Given the upload is skynode specific, and has to be for the foreseeable future, your proposal to use `make px4_fmu-v5x upload_skynode_usb` and `make px4_fmu-v5x upload_skynode_wifi` make a lot of sense.

hamishwillee commented:
Any guidance on using `remote_update_fmu.sh`?

hamishwillee commented:
Docs created "for discussion" in https://github.com/PX4/PX4-user_guide/pull/2631

MaEtUgR commented:
I agree that it's much clearer if the name reflects that it's a Skynode-specific upload and adjusted accordingly. I also double checked it works like expected:

```
$ make px4_fmu-v5x upload_skynode_usb
-- PX4 version: v1.14.0-beta2-356-g5d7eac91e0 (1.14.0)
-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.10.6", minimum required is "3") 
-- PX4 config file: /home/maetugr/PX4-Autopilot2/boards/px4/fmu-v5x/default.px4board

...

[1190/1193] Linking CXX executable px4_fmu-v5x_default.elf
Memory region         Used Size  Region Size  %age Used
      FLASH_ITCM:          0 GB      2016 KB      0.00%
      FLASH_AXIM:     2050137 B      2016 KB     99.31%
        ITCM_RAM:          0 GB        16 KB      0.00%
        DTCM_RAM:          0 GB       128 KB      0.00%
           SRAM1:       92508 B       368 KB     24.55%
           SRAM2:          0 GB        16 KB      0.00%
[1192/1193] Uploading PX4
Uploading to 10.41.1.1...
firmware.px4
auterion@10.41.1.1's password: 
update-dev.tar                                                                                                                                                                                                                                                             100% 1850KB  17.4MB/s   00:00    
auterion@10.41.1.1's password: 
Waiting for flight controller state...
DISARMED, continuing
Updating the FMU firmware...
Waiting for bootloader...

Found board id: 51,0 bootloader version: 5 on /dev/ttymxc3
Loaded firmware for board id: 51,0 size: 2050137 bytes (99.31%) 

sn: 2a0027002a002700185138342a0027001851383439353636
chip: 10016451
family: b'STM32F7[6|7]x'
revision: b'Z'
flash: 2064384 bytes
Windowed mode: False

Erase  : [====================] 100.0%
Program: [====================] 100.0%
Verify : [====================] 100.0%
Rebooting. Elapsed Time 41.912

Connection to 10.41.1.1 closed.
```

hamishwillee commented:
Looks good to me. Who in the technical team do we get to review/approve this?

MaEtUgR commented:
@hamishwillee I rebased on latest main and added the possibility to revert the autopilot to the originally deployed release version using:
`./Tools/auterion/upload_skynode.sh --revert`

hamishwillee reviewed (approved):
@bkueng I have reviewed this from a usability point of view . The scripts look sane, but I haven't run them and I'm not expert enough to be sure. Can you review/merge or suggest someone else who could do this?

bkueng commented on Tools/auterion/remote_update_fmu.sh:
```diff
@@ -50,53 +50,66 @@ while getopts ":f:c:d:p:u:" opt; do
-rm -rf "$tmp_dir"
 
-# grab status output
+# grab status output for flshing progress
```
flashing

MaEtUgR replied:
Thanks for spotting 🤦‍♂️ Corrected in https://github.com/PX4/PX4-Autopilot/commit/cdc784eeb3b5257fcb6f38406705224748b88dbd

bkueng commented on Tools/auterion/remote_update_fmu.sh:
```diff
@@ -50,53 +50,66 @@ while getopts ":f:c:d:p:u:" opt; do
-firmware_path=""
+if [ "$revert" == true ]; then
+	# revert to the release version which was originally deployed
+	cmd="cp /persistent/shared_container_dir/fmu/update.tar /persistent/shared_container_dir/fmu/update-dev.tar"
```
Can you use `target_dir=/shared_container_dir/fmu` from below?

MaEtUgR replied:
How's the mounting? Does it also work with `/shared_container_dir/fmu` instead of `/persistent/shared_container_dir/fmu`. Looking at the SCP command it should but I haven't checked 🤔

MaEtUgR replied:
I reused the variable like in the scp command for flashing without the persistent root folder in https://github.com/PX4/PX4-Autopilot/commit/cdc784eeb3b5257fcb6f38406705224748b88dbd. Echoing the output it's now:
`cp /shared_container_dir/fmu/update.tar /shared_container_dir/fmu/update-dev.tar`
but I can't test since I'm missing the hardware now. Could someone give this command a try? 😇

bkueng commented on Tools/auterion/upload_skynode.sh:
```diff
@@ -29,8 +37,15 @@ done
 [ -z "$AUTOPILOT_USER" ] && AUTOPILOT_USER=$DEFAULT_AUTOPILOT_USER
 
-echo "Uploading to $AUTOPILOT_HOST..."
+ARGUMENTS=""
```
Can you use an array to handle spaces in the path correctly?
```
array=()
array+=(xy)
```

Then pass it as `"${array[@]}"`

MaEtUgR replied:
Thanks for the hint. I fighted with the spaces and didn't know about this possibility. Done in cdc784e
When echoing commands it still looks correct.

hamishwillee commented:
@bkueng Looks like it needs a re-review and possibly a retest.

bkueng commented on Tools/auterion/upload_skynode.sh:
```diff
@@ -0,0 +1,51 @@
+        DEFAULT_AUTOPILOT_USER="${i#*=}"
+        ;;
+        --revert)
+        REVERT_AUTOPILOT_ARGUMENT=" -r"
```
```suggestion
        REVERT_AUTOPILOT_ARGUMENT=-r
```

MaEtUgR replied:
Added to https://github.com/PX4/PX4-Autopilot/commit/b9c4dc8b97c81af9fdd2bea41554a33f3e3cd561

bkueng commented on Tools/auterion/upload_skynode.sh:
```diff
@@ -0,0 +1,51 @@
+
+echo "Flashing $AUTOPILOT_HOST ..."
+
+"$DIR"/remote_update_fmu.sh ${ARGUMENTS[@]}
```
```suggestion
"$DIR"/remote_update_fmu.sh "${ARGUMENTS[@]}"
```

MaEtUgR replied:
I add this to b9c4dc8b97c81af9fdd2bea41554a33f3e3cd561.

Note that I left away all the quotes of the individual parts e.g. not doing `ARGUMENTS+=(${PX4_BINARY_FILE:+"-f $PX4_BINARY_FILE"})`. It works either with the individual parts or the entire argument list quoted but when doing both it just skips through flashing.
```
make px4_fmu-v5x upload_skynode_usb        
[0/1] Uploading PX4
Flashing 10.41.1.1 ...
```
Didn't invest more time to figure out why.

bkueng commented on Tools/auterion/upload_skynode.sh:
```diff
@@ -0,0 +1,51 @@
+ARGUMENTS+=(-d $AUTOPILOT_HOST)
+ARGUMENTS+=(-p $AUTOPILOT_PORT)
+ARGUMENTS+=(-u $AUTOPILOT_USER)
+ARGUMENTS+=($PX4_BINARY_FILE_ARGUMENT)
```
This is still incorrect with spaces. You need to use `ARGUMENTS+=(-f "$PX4_BINARY_FILE")`

MaEtUgR replied:
Well the file argument is optional because it's not needed when reverting. I can of course fill the variable with the file name only and then conditionally also add the -f before if the name is not empty. I thought it's cleaner that way but will check this option 🙏

MaEtUgR replied:
I found https://stackoverflow.com/a/20306982/6326048 and learned by using `:+` I can make an entire part optional based on if a variable is set. It also works on zsh. I added this to b9c4dc8b97c81af9fdd2bea41554a33f3e3cd561.

MaEtUgR commented:
I just tested this again with the latest changes. I squashed them together with the other review addressing since I wasn't even able to verify if the previous changes were working.

bkueng commented on boards/px4/fmu-v5x/cmake/upload.cmake:
Can you also add this to v6x?

MaEtUgR replied:
Added in 8913b4e4861c437b30b1fa16f5cf278a149e1937

bkueng commented on Tools/auterion/upload_skynode.sh:
```diff
@@ -0,0 +1,51 @@
+ARGUMENTS+=(-d $AUTOPILOT_HOST)
+ARGUMENTS+=(-p $AUTOPILOT_PORT)
+ARGUMENTS+=(-u $AUTOPILOT_USER)
+ARGUMENTS+=(${PX4_BINARY_FILE:+-f $PX4_BINARY_FILE})
```
```suggestion
ARGUMENTS+=(${PX4_BINARY_FILE:+-f "$PX4_BINARY_FILE"})
```
It's still not correct:
```
x='a b'
a=()
a+=(${x:+-f $x})
for i in "${a[@]}"; do echo "--$i--"; done
---f--
--a--
--b--
```

MaEtUgR replied:
Ah spaces in the path would not be handled correctly 💡 Thanks for that test output script idea and the clarifications in person. I corrected in the same commit again and tested against the hardware, just to be sure.
